### PR TITLE
Move ChildActivity handling into base Activity class

### PR DESCRIPTION
**Pull request by tovl**
Depends on #16369 and #16382.

With all activities now converted to the `ChildActivity` paradigm, some common patterns can be deduplicated by letting `Activity.TickOuter` handle childactivities.

### CODE REVIEW

File: OpenRA.Game/Activities/Activity.cs [hunk 1]
```diff
@@ -41,11 +41,13 @@ protected Activity ChildActivity
 		public Activity NextActivity { get; protected set; }
 
 		public bool IsInterruptible { get; protected set; }
+		public bool ChildHasPriority { get; protected set; }
 		public bool IsCanceling { get { return State == ActivityState.Canceling; } }
 
 		public Activity()
 		{
 			IsInterruptible = true;
+			ChildHasPriority = true;
```
Comment (pchote) [inline OpenRA.Game/Activities/Activity.cs]: Having this as a flag doesn't feel right to me. Is there a specific reason you didn't implement this using a virtual method that the special cases can override?

Comment (tovl) [reply OpenRA.Game/Activities/Activity.cs]: The idea is that there shouldn't be special cases: Either the child has priority over the parent or the parent has priority over the child. I've now refactored `Enter` to also fit into this paradigm (all other activities already did). This means that only the base `Activity` class needs to set `ChildActivity` directly and the setter can be made private.

**Comment by reaperrr** [comment]
#16382 was merged.

**Comment by tovl** [comment]
Rebased.

**Comment by tovl** [comment]
Added a default `Tick` method to `Activity` that will immediately return `NextActivity` for activities that only need to run for one tick. This would make replacing instances of `CallFunc` with full activities slightly less verbose.

**Comment by tovl** [comment]
Rebased and removed the `Done` state. Once the state is set to `Done`, that activity will never be referenced to again because all instances of activities queueing themselves have already been removed.

File: OpenRA.Game/Activities/Activity.cs [hunk 2]
```diff
@@ -31,45 +31,56 @@ public abstract class Activity
...
 		}
 
 		public Activity TickOuter(Actor self)
 		{
-			if (State == ActivityState.Done && Game.Settings.Debug.StrictActivityChecking)
```
Comment (pchote) [inline OpenRA.Game/Activities/Activity.cs]: This was the only place that used `StrictActivityChecking`, so that can now be removed from `DebugSettings` and the advanced settings UI.

File: OpenRA.Game/Activities/Activity.cs [hunk 3]
```diff
@@ -31,45 +31,56 @@ public abstract class Activity
...
+			{
+				ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
+				if (ChildActivity != null)
+					return this;
+			}
+
 			var ret = Tick(self);
-			if (ret != this)
+
+			if (ChildActivity != null && (ChildActivity.State == ActivityState.Queued || !ChildHasPriority))
```
Comment (pchote) [inline OpenRA.Game/Activities/Activity.cs]: Can you please add a comment here explaining that we want to tick the child activity if it was just queued to avoid a single-tick delay where the actor appears to stall without doing anything?

Comment (pchote) [reply OpenRA.Game/Activities/Activity.cs]: I assume that we have already reached the stage where we would never not want to pre-tick the child, so that it is fine to not expose a way to not do this.

File: OpenRA.Game/Activities/Activity.cs [hunk 4]
```diff
@@ -31,45 +31,56 @@ public abstract class Activity
...
 			{
-				State = ActivityState.Done;
-				OnLastRun(self);
+				ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
+				if (!ChildHasPriority && ChildActivity == null)
```
Comment (pchote) [inline OpenRA.Game/Activities/Activity.cs]: Can you please explain this? It doesn't make sense to me that a child activity completing should force the parent (which may still have work to do?) to be skipped without cancellation.

**Comment by pchote** [comment]
> Once the state is set to Done, that activity will never be referenced to again because all instances of activities queueing themselves have already been removed.

Isn't this only true because we've made an effort to fix the bogus cases that were previously in the code? I'd be a lot happier if we could guarantee that we (or modders) can't accidentally break this again in the future, by keeping the state and hardcoding the `StrictActivityChecking` behaviour to true.

**Comment by obrakmann** [comment]
> removed the Done state. Once the state is set to Done, that activity will never be referenced to again because all instances of activities queueing themselves have already been removed.

Just for the record, I don't think this is a good idea. All current instances may have been fixed, but that doesn't mean that someone in the future (or an external modder) won't come along and introduce the mistake in another activity. Also it costs basically nothing at all.

File: OpenRA.Game/Activities/Activity.cs (hunk 4 above)
Comment (tovl) [reply OpenRA.Game/Activities/Activity.cs]: `ChildHasPriority = false` is meant for activities that need to calculate stuff before the childactivity ticks _as long as there are children running_: Only `MoveAdjacentTo` and `AttackMoveActivity` need this. The other activities are happy to let the child finish before doing their own logic.

Perhaps it would be clearer to give the flag another name (although I'm not sure what that should be)?

File: OpenRA.Game/Activities/Activity.cs (hunk 2 above)
Comment (tovl) [reply OpenRA.Game/Activities/Activity.cs]: Done.

File: OpenRA.Game/Activities/Activity.cs (hunk 3 above)
Comment (tovl) [reply OpenRA.Game/Activities/Activity.cs]: Done.

All instances where we did not pretick a childactivity where because it was already queued in `Activity.OnFirstRun`, whenever it was queued in `Activity.Tick` we did do a pre-tick. So automating this pattern by letting `TickOuter` figure it out makes sense to me.

**Review by tovl — COMMENTED** [review]
Yeah, that seems like a valid concern. Reverted and changed to always strictly check instead.

File: OpenRA.Game/Activities/Activity.cs (hunk 4 above)
Comment (pchote) [reply OpenRA.Game/Activities/Activity.cs]: This feels a bit too magic for me - I can see us running into a brick wall if we find the need for a third type of activity in the future. I'm all for the current behavior with `ChildHasPriority = true`, but I'm still thinking that it would be simpler and more robust to make this flag completely disable the child `RunActivity` when disabled.  We could then expose a `TickChild` method that activities like `AttackMoveActivity` can call when when better suits them (and throws an exception if it is called by a method that hasn't opted out of the automatic behaviour). This would make the behaviour much more obvious.

**Comment by pchote** [comment]
Do we have any cases left where an activity returns something that isn't `this` or `NextActivity`?

If not, would it make sense to change `Tick` to return a bool (true = next activity, false = not finished yet) instead? This would make it harder to violate the "Things to be aware of" points noted at the top of the class.

**Comment by tovl** [comment]
I've made `Tick` return bool instead of `Activity`. Also, since we always make sure that childactivities finish before going to `NextActivity` (and we should), I've build this into `TickOuter`. This can be used to simplify some cases where we set a boolean immediately after queueing a childactivity to indicate we are done after that. Also `ReturnToBase` is now less convoluted to cancel.

File: OpenRA.Game/Activities/Activity.cs (hunk 4 above)
Comment (tovl) [reply OpenRA.Game/Activities/Activity.cs]: OK. `AttackMoveActivity` and `MoveAdjacentTo` now call `TickChild` themselves.

File: OpenRA.Game/Activities/Activity.cs [hunk 5]
```diff
@@ -59,17 +56,49 @@ public Activity TickOuter(Actor self)
...
+		protected bool TickChild(Actor self)
+		{
+			ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
+			if (ChildActivity != null)
+				return false;
+
+			return true;
+		}
+
+		/// <summary>
```
Comment (pchote) [inline OpenRA.Game/Activities/Activity.cs]: It would be good if we could add some more detail here.
Maybe something like:
```xml
<summary>
Called every tick to run activity logic. Returns false if the activity should
remain active, or true if it is complete. Cancelled activities must ensure they
return the actor to a consistent state before returning true.

Child activities can be queued using QueueChild, and these will be ticked
instead of the parent while they are active. Activities that need to run logic
in parallel with child activities should set ChildHasPriority to false and
manually call TickChildren.

Queuing one or more child activities and returning true is valid, and causes
the activity to be completed immediately (without ticking again) once the
children have completed.
</summary>
```
?

File: OpenRA.Game/Activities/Activity.cs [hunk 6]
```diff
@@ -59,17 +56,49 @@ public Activity TickOuter(Actor self)
...
 
-			return ret;
+			return this;
 		}
 
-		public abstract Activity Tick(Actor self);
+		protected bool TickChild(Actor self)
+		{
+			ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
+			if (ChildActivity != null)
```
[the remark this answers is not in the document]
Comment (tovl) [reply OpenRA.Game/Activities/Activity.cs]: Done.

File: OpenRA.Game/Activities/Activity.cs (hunk 5 above)
Comment (tovl) [reply OpenRA.Game/Activities/Activity.cs]: Done.

**Comment by pchote** [comment]
Needs a rebase after #16365.

**Comment by tovl** [comment]
Rebased.

**Comment by matjaeck** [comment]
Something is wrong with the settings it seems, the game crashes when you try to open them.

<details>

```
OpenRA engine version git-81fa33e47
Red Alert mod version git-81fa33e47
on map 0c119c6c942292af5aa1899f29d94eb7a60c791f (Desert Shellmap by Scott_NZ).
Date: 2019-05-20 16:01:47Z
Operating System: Linux (Unix 4.15.0.47)
Runtime Version: Mono 5.20.1.19 (tarball Thu Apr 11 09:02:17 UTC 2019) CLR 4.0.30319.42000
Exception of type `System.Reflection.TargetInvocationException`: Exception has been thrown by the target of an invocation.
Inner
  Exception of type `System.InvalidOperationException`: DebugSettings does not contain a preference type StrictActivityChecking
    at (wrapper managed-to-native) System.Reflection.MonoCMethod.InternalInvoke(System.Reflection.MonoCMethod,object,object[],System.Exception&)
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters, System.Boolean wrapExceptions) [0x00005] in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters, System.Boolean wrapExceptions) [0x0001a] in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Reflection.MonoCMethod.DoInvoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) <0x7f0451e78ff0 + 0x000b3> in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Reflection.MonoCMethod.Invoke (System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) <0x7f0451e79130 + 0x0003c> in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Reflection.ConstructorInfo.Invoke (System.Object[] parameters) <0x7f0451e654b0 + 0x00034> in <6649516e5b3542319fb262b421af0adb>:0 
  at OpenRA.ObjectCreator.CreateUsingArgs (System.Reflection.ConstructorInfo ctor, System.Collections.Generic.Dictionary`2[TKey,TValue] args) [0x00054] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.ObjectCreator.CreateObject[T] (System.String className, System.Collections.Generic.Dictionary`2[TKey,TValue] args) [0x0004e] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.Widgets.Widget+<>c__DisplayClass27_0.<PostInit>b__0 (System.String l) [0x00000] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at System.Linq.Enumerable+SelectArrayIterator`2[TSource,TResult].ToArray () [0x00012] in <35ad2ebb203f4577b22a9d30eca3ec1f>:0 
  at System.Linq.Enumerable.ToArray[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0001f] in <35ad2ebb203f4577b22a9d30eca3ec1f>:0 
  at OpenRA.Widgets.Widget.PostInit (OpenRA.Widgets.WidgetArgs args) [0x00044] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, OpenRA.MiniYamlNode node) [0x001b3] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, System.String w) [0x0002a] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.Widgets.Ui.OpenWindow (System.String id, OpenRA.Widgets.WidgetArgs args) [0x00000] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.Game.OpenWindow (System.String widget, OpenRA.Widgets.WidgetArgs args) [0x0003c] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.Mods.Common.Widgets.Logic.MainMenuLogic+<>c__DisplayClass13_0.<.ctor>b__3 () [0x00040] in <c66803f6888248c0a71c764d97c1036b>:0 
  at OpenRA.Mods.Common.Widgets.ButtonWidget.<.ctor>b__42_5 (OpenRA.MouseInput _) [0x00000] in <c66803f6888248c0a71c764d97c1036b>:0 
  at OpenRA.Mods.Common.Widgets.ButtonWidget.HandleMouseInput (OpenRA.MouseInput mi) [0x00076] in <c66803f6888248c0a71c764d97c1036b>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x000a0] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.Widgets.Ui.HandleInput (OpenRA.MouseInput mi) [0x0001e] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.DefaultInputHandler+<>c__DisplayClass5_0.<OnMouseInput>b__0 () [0x00000] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.Sync.RunUnsynced[T] (System.Boolean checkSyncHash, OpenRA.World world, System.Func`1[TResult] fn) [0x00023] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.DefaultInputHandler.OnMouseInput (OpenRA.MouseInput input) [0x0000d] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.Platforms.Default.Sdl2Input.PumpInput (OpenRA.Platforms.Default.Sdl2PlatformWindow device, OpenRA.IInputHandler inputHandler) [0x001d4] in <2ef045f9f8ec4a299f7167ae8d373381>:0 
  at OpenRA.Platforms.Default.Sdl2PlatformWindow.PumpInput (OpenRA.IInputHandler inputHandler) [0x00006] in <2ef045f9f8ec4a299f7167ae8d373381>:0 
  at OpenRA.Renderer.EndFrame (OpenRA.IInputHandler inputHandler) [0x0000c] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.Game.RenderTick () [0x0014b] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.Game.Loop () [0x0014d] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.Game.Run () [0x0003c] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00010] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
  at OpenRA.Program.Main (System.String[] args) [0x00044] in <a8d01b0d215f4bcb96e734d4edddcbb3>:0 
```

**Review by ghost — CHANGES_REQUESTED** [review]
I haven't tested every activity yet, and so far I only noticed the following issues:

- idle yaks will not respond to queued orders (we had this issue before)
- fully loaded harvesters sometimes won't unload their resources and block the refinery. This happens in AI games frequently after some minutes at many refineries with sometimes only one or two harvesters (tested on tournament island, 4 AI, fastest). This specific issue seems to be a regression compared to bleed.
   <details>
   I don't know the exact reproduction, but it happens frequently (also when not floating money).

    harvfail

**Comment by tovl** [comment]
Fixed the settings issue.

**Comment by tovl** [comment]
> idle yaks will not respond to queued orders (we had this issue before)

This issue is present in bleed as well and was regressed by #16365. the problem is that `LandWhenIdle` defaults to true even for aircraft that have no valid terrain for landing, so the aircraft gets stuck in a holding pattern until canceled. This should be handled in a separate fixup PR.

**Comment by tovl** [comment]
The harvester issue should be fixed now as well

**Comment by matjaeck** [comment]
Can confirm the fixes. Didn't notice any other issues, but haven't tested everything, so can only give a tentative :+1: for now.

**Review by ghost — CHANGES_REQUESTED** [review]
Planes or helicopters will not attack when queuing a force attack order after move orders. It looks like they're aiming but something prevents them from firing. Queuing "normal" attacks works.

**Comment by tovl** [comment]
This bug is present in bleed as well, so is not relevant for this PR.

**Comment by tovl** [comment]
Rebased.

**Review by ghost — DISMISSED** [review]
Tested in RA.

**Comment by pchote** [comment]
Testing in #16584 uncovered a fairly nasty edge case in the activity state plumbing. I'm not sure whether it makes sense to fix it here, or in its own PR.

If you queue a child activity and then cancel it before it can tick it will go directly from `ActivityState.Queued` to `ActivityState.Canceling`. This allows it to tick without `OnFirstRun` ever being called, which breaks a lot of assumptions!

**Comment by pchote** [comment]
Also needs a rebase, but i'm assuming that we're all on the same page about sorting out the aircraft activity snarl before these other PRs.

File: OpenRA.Game/Activities/Activity.cs [hunk 7]
```diff
@@ -59,17 +56,56 @@ public Activity TickOuter(Actor self)
...
+
+		protected bool TickChild(Actor self)
+		{
+			ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
+			return ChildActivity == null;
 		}
 
-		public abstract Activity Tick(Actor self);
+		/// <summary>
+		/// Called every tick to run activity logic.Returns false if the activity should
```
Comment (pchote) [inline OpenRA.Game/Activities/Activity.cs]: Missing space between `logic.Returns`

File: OpenRA.Game/Activities/Activity.cs [hunk 8]
```diff
@@ -59,17 +56,56 @@ public Activity TickOuter(Actor self)
...
+		/// remain active, or true if it is complete.Cancelled activities must ensure they
```
Comment (pchote) [inline OpenRA.Game/Activities/Activity.cs]: Another missing space between `complete.Cancelled`

File: OpenRA.Game/Activities/Activity.cs [hunk 9]
```diff
@@ -59,17 +56,56 @@ public Activity TickOuter(Actor self)
...
+		/// return the actor to a consistent state before returning true.
+		///
+		/// Child activities can be queued using QueueChild, and these will be ticked
+		/// instead of the parent while they are active.Activities that need to run logic
```
Comment (pchote) [inline OpenRA.Game/Activities/Activity.cs]: Here too

**Review by pchote — COMMENTED** [review]
It would be helpful for reviewing and testing if we could split the `Enter`- and `Move`-related changes into their own PRs (and I assume the `Aircraft` changes will be dealt with by the current aircraft PRs) so that this one is just the plumbing changes.

@abcdefg30 hit the nail on the head by pointing out that the activity PRs have been so painful because they are doing so much at once - in some cases this has been unavoidable, but this is one case where we can.

File: OpenRA.Game/Activities/Activity.cs (hunk 7 above)
Comment (tovl) [reply OpenRA.Game/Activities/Activity.cs]: Fixed.

File: OpenRA.Game/Activities/Activity.cs (hunk 8 above)
Comment (tovl) [reply OpenRA.Game/Activities/Activity.cs]: Fixed.

File: OpenRA.Game/Activities/Activity.cs (hunk 9 above)
Comment (tovl) [reply OpenRA.Game/Activities/Activity.cs]: Fixed.

**Comment by tovl** [comment]
Removed the `Enter` refactor and rebased.

**Comment by teinarss** [comment]
Needs a new rebase now :(

**Comment by reaperrr** [comment]
Let's wait with that until #16721 and #16509 are merged.

**Comment by pchote** [comment]
We need to branch a new playtest series ASAP, so lets defer this to Next + 1.

**Comment by tovl** [comment]
Let's not. Leaving this for next +1 will leave the general activity refactor in a half-finished state. I'd rather finish this now and be done with it. In fact, this has been as good as done for a while now and apart from yet another rebase there is nothing left to do.

**Comment by tovl** [comment]
Rebased. 

> Let's wait with that until #16721 and #16509 are merged.

Let's please give priority to this PR. Every new PR that touches any part of the activity code in some way will necessitate another rebase and that's just a waste of time. If we wait till after the playtest I might as well start from scratch.

**Comment by pchote** [comment]
I'd like to get #16509 in first because that is the bottleneck blocking the other aircraft and balance PRs, but then ok - lets get this in if we don't find any major regressions afterwards.

**Comment by reaperrr** [comment]
#16509 was merged.

**Comment by tovl** [comment]
Rebased.

File: OpenRA.Mods.Common/Activities/Move/VisualMoveIntoTarget.cs [hunk 10]
```diff
@@ -35,45 +35,41 @@ protected override void OnFirstRun(Actor self)
...
 			var delta = targetPos - currentPos;
 			var facing = delta.HorizontalLengthSquared != 0 ? delta.Yaw.Facing : mobile.Facing;
 			if (facing != mobile.Facing)
 			{
-				var turn = ActivityUtils.RunActivity(self, new Turn(self, facing));
-				if (turn != null)
-					QueueChild(turn);
-
-				return this;
+				mobile.Facing = Util.TickFacing(mobile.Facing, facing, mobile.TurnSpeed);
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Move/VisualMoveIntoTarget.cs]: Why not queue the ChildActivity?

Again, would be ideal if the logical changes can be kept in a separate commit from the mechanical ones.

File: OpenRA.Mods.Common/Activities/RemoveSelf.cs [hunk 11]
```diff
@@ -15,11 +15,12 @@ namespace OpenRA.Mods.Common.Activities
...
 	public class RemoveSelf : Activity
 	{
-		public override Activity Tick(Actor self)
+		public override bool Tick(Actor self)
 		{
-			if (IsCanceling) return NextActivity;
+			if (IsCanceling) return true;
 			self.Dispose();
-			return null;
+			Cancel(self);
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/RemoveSelf.cs]: What does cancelling achieve here?

File: OpenRA.Mods.Common/Activities/Transform.cs [hunk 12]
```diff
@@ -168,11 +168,5 @@ public Order IssueOrderForTransformedActor(Actor newActor)
 		{
 			return new Order(orderString, newActor, target, true);
 		}
-
-		public override Activity Tick(Actor self)
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Transform.cs]: This is in the wrong commit ("Make Tick return bool." instead of "Add default for Activity-Tick").

File: OpenRA.Mods.Common/Activities/UnloadCargo.cs [hunk 13]
```diff
@@ -124,15 +124,16 @@ public override Activity Tick(Actor self)
 
 			if (!unloadAll || !cargo.CanUnload())
 			{
-				Cancel(self, true);
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/UnloadCargo.cs]: What does removing this cancel achieve?

File: OpenRA.Mods.Common/Traits/Buildings/TransformsIntoAircraft.cs [hunk 14]
```diff
@@ -130,7 +130,7 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 			if (!order.Queued && activity.NextActivity != null)
 				activity.NextActivity.Cancel(self);
 
-			activity.Queue(self, new IssueOrderAfterTransform(order.OrderString, order.Target));
+			activity.Queue(new IssueOrderAfterTransform(order.OrderString, order.Target));
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/TransformsIntoAircraft.cs]: This is in the wrong commit ("Make Tick return bool." instead of "Move ChildActivity handling into base Activity class.").

File: OpenRA.Mods.Common/Traits/Buildings/TransformsIntoEntersTunnels.cs [hunk 15]
```diff
@@ -100,7 +100,7 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 			if (!order.Queued && activity.NextActivity != null)
 				activity.NextActivity.Cancel(self);
 
-			activity.Queue(self, new IssueOrderAfterTransform(order.OrderString, order.Target));
+			activity.Queue(new IssueOrderAfterTransform(order.OrderString, order.Target));
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/TransformsIntoEntersTunnels.cs]: This is in the wrong commit ("Make Tick return bool." instead of "Move ChildActivity handling into base Activity class.").

File: OpenRA.Mods.Common/Traits/Buildings/TransformsIntoMobile.cs [hunk 16]
```diff
@@ -111,7 +111,7 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 				if (!order.Queued && activity.NextActivity != null)
 					activity.NextActivity.Cancel(self);
 
-				activity.Queue(self, new IssueOrderAfterTransform("Move", order.Target));
+				activity.Queue(new IssueOrderAfterTransform("Move", order.Target));
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/TransformsIntoMobile.cs]: This is in the wrong commit ("Make Tick return bool." instead of "Move ChildActivity handling into base Activity class.").

File: OpenRA.Mods.Common/Traits/Buildings/TransformsIntoPassenger.cs [hunk 17]
```diff
@@ -124,7 +124,7 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 			if (!order.Queued && activity.NextActivity != null)
 				activity.NextActivity.Cancel(self);
 
-			activity.Queue(self, new IssueOrderAfterTransform(order.OrderString, order.Target));
+			activity.Queue(new IssueOrderAfterTransform(order.OrderString, order.Target));
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/TransformsIntoPassenger.cs]: This is in the wrong commit ("Make Tick return bool." instead of "Move ChildActivity handling into base Activity class.").

File: OpenRA.Mods.Common/Traits/Buildings/TransformsIntoRepairable.cs [hunk 18]
```diff
@@ -118,7 +118,7 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 			if (!order.Queued && activity.NextActivity != null)
 				activity.NextActivity.Cancel(self);
 
-			activity.Queue(self, new IssueOrderAfterTransform(order.OrderString, order.Target));
+			activity.Queue(new IssueOrderAfterTransform(order.OrderString, order.Target));
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/TransformsIntoRepairable.cs]: This is in the wrong commit ("Make Tick return bool." instead of "Move ChildActivity handling into base Activity class.").

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeploy.cs [hunk 19]
```diff
@@ -132,7 +132,7 @@ Activity IWrapMove.WrapMove(Activity moveInner)
 				return moveInner;
 
 			var activity = new DeployForGrantedCondition(self, this, true);
-			activity.Queue(self, moveInner);
+			activity.Queue(moveInner);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeploy.cs]: This is in the wrong commit ("Make Tick return bool." instead of "Move ChildActivity handling into base Activity class.").

File: OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs [hunk 20]
```diff
@@ -106,35 +109,21 @@ public override Activity Tick(Actor self)
 			// Target is hidden or dead, and we don't have a fallback position to move towards
 			var noTarget = useLastVisibleTarget && !lastVisibleTarget.IsValidFor(self);
 
-			// Inner move order has completed.
```
[the remark this answers is not in the document]
Comment (pchote) [reply OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs]: (already fixed while I was working on the rest of the review)

File: OpenRA.Mods.Common/Activities/WaitForTransport.cs [hunk 21]
```diff
@@ -11,29 +11,27 @@
...
 	{
 		readonly ICallForTransport transportable;
+		readonly Activity inner;
 
-		public WaitForTransport(Actor self, Activity innerActivity)
+		public WaitForTransport(Actor self, Activity inner)
 		{
 			transportable = self.TraitOrDefault<ICallForTransport>();
-			QueueChild(self, innerActivity);
+			this.inner = inner;
```
[the remark this answers is not in the document]
Comment (pchote) [reply OpenRA.Mods.Common/Activities/WaitForTransport.cs]: (already fixed while I was working on the rest of the review)

File: OpenRA.Mods.Common/Activities/RemoveSelf.cs (hunk 11 above)
Comment (tovl) [reply OpenRA.Mods.Common/Activities/RemoveSelf.cs]: It removes the rest of the activity queue, so it achieves the same effect as returning `null` would have before.

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs [hunk 22]
```diff
@@ -108,14 +108,14 @@ public override Activity Tick(Actor self)
...
+				return true;
 			}
 
 			// If all valid weapons have depleted their ammo and Rearmable trait exists, return to RearmActor to reload and then resume the activity
 			if (rearmable != null && !useLastVisibleTarget && attackAircraft.Armaments.All(x => x.IsTraitPaused || !x.Weapon.IsValidAgainst(target, self.World, self)))
 			{
-				QueueChild(new ReturnToBase(self, aircraft.Info.AbortOnResupply));
-				return this;
+				QueueChild(new ReturnToBase(self));
+				return aircraft.Info.AbortOnResupply;
```
[the remark this answers is not in the document]
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: This uncomplicates the `AbortOnResupply` behaviour by removing it from `ReturnToBase` and handling it here.

File: OpenRA.Mods.Common/Activities/Move/VisualMoveIntoTarget.cs (hunk 10 above)
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Move/VisualMoveIntoTarget.cs]: It's to prevent the turn from overshooting.

File: OpenRA.Mods.Common/Activities/UnloadCargo.cs (hunk 13 above)
Comment (tovl) [reply OpenRA.Mods.Common/Activities/UnloadCargo.cs]: The same effect can be achieved by returning `true` instead. So this is simply to make it consistent with the other changes.

File: mods/common/chrome/settings.yaml [hunk 23]
```diff
@@ -669,11 +669,3 @@ Background@SETTINGS_PANEL:
 							Height: 20
 							Font: Regular
 							Text: Enable Debug Commands in Replays
-						Checkbox@STRICTACTIVITY_CHECKBOX:
```
[the remark this answers is not in the document]
Comment (tovl) [reply mods/common/chrome/settings.yaml]: Done.

File: OpenRA.Game/Activities/Activity.cs [hunk 24]
```diff
@@ -55,33 +56,58 @@ public Activity TickOuter(Actor self)
...
+				finishing |= lastRun;
 			}
 
-			var ret = Tick(self);
+			// The parent determines whether the child gets a chance at ticking.
+			else
+				lastRun = Tick(self);
 
+			// Avoid a single tick delay if the childactivity was just queued.
 			if (ChildActivity != null && ChildActivity.State == ActivityState.Queued)
```
[the remark this answers is not in the document]
Comment (tovl) [reply OpenRA.Game/Activities/Activity.cs]: Done.

File: OpenRA.Mods.Cnc/Traits/Attack/AttackTDGunboatTurreted.cs [hunk 25]
```diff
@@ -46,27 +46,27 @@ public AttackTDGunboatTurretedActivity(Actor self, Target target, bool allowMove
...
 					if (hasTicked && attack.RequestedTarget.Type == TargetType.Invalid)
-						return NextActivity;
+						return true;
 
 					attack.SetRequestedTarget(self, target);
 					hasTicked = true;
 				}
 
-				return NextActivity;
+				return false;
```
[the remark this answers is not in the document]
Comment (tovl) [reply OpenRA.Mods.Cnc/Traits/Attack/AttackTDGunboatTurreted.cs]: Fixed.

**Comment by tovl** [comment]
Split out the logic changes into separate commits.

**Review by pchote — DISMISSED** [review]

**Comment by tovl** [comment]
Rebased.

**Review by teinarss — APPROVED** [review]
LGTM

**Comment by teinarss** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/a0d93ae3c83f7f105056e47817db76ecd260f16d...83c92b3f8dac5c41e8c8b36aa92ec158cd25448b
